File: Methods/Methods-Python/cosineDistance.py
import numpy as np


def consineDistance(p1, p2):
    dotProduct = 0
    lengthSquaredp1 = 0
    lengthSquaredp2 = 0

    for i in range(0, len(p1)):
        lengthSquaredp1 += p1[i] * p1[i]
        lengthSquaredp2 += p2[i] * p2[i]
        dotProduct += p1[i] * p2[i]

    denominator = np.sqrt(lengthSquaredp1) * np.sqrt(lengthSquaredp2)

    if denominator < dotProduct:
        denominator = dotProduct

    if denominator == 0 and dotProduct == 0:
        return 0

    return 1 - dotProduct / denominator

File: Methods/Methods-Java/Java source-code/cosineDistance.java


public class cosineDistance {
    public static double cosineDistance_m(Double[] p1, Double[] p2) {
        double dotProduct = 0.0;
        double lengthSquaredp1 = 0.0;
        double lengthSquaredp2 = 0.0;
        for (int i = 0; i < p1.length; i++) {
            lengthSquaredp1 += p1[i] * p1[i];
            lengthSquaredp2 += p2[i] * p2[i];
            dotProduct += p1[i] * p2[i];
        }
        double denominator = Math.sqrt(lengthSquaredp1) * Math.sqrt(lengthSquaredp2);
        if (denominator < dotProduct) {
            denominator = dotProduct;
        }
        if (denominator == 0 && dotProduct == 0) {
            return 0;
        }
        return 1.0 - dotProduct / denominator;
    }
}